src/hotspot/share/nmt/vmatree.cpp

```diff
@@ -215,3 +216,63 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
   }
   return diff;
 }
+
+#ifdef ASSERT
+void VMATree::print_on(outputStream* out) {
+  visit_in_order([&](TreapNode* current) {
+    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(out_state(current).mem_tag()),
+               statetype_to_string(out_state(current).type()));
+  });
+  out->cr();
+}
+#endif
+
+VMATree::SummaryDiff VMATree::set_tag(position from, size size, MemTag tag) {
+  VMATreap::Range range = _tree.find_enclosing_range(from);
+  assert(range.start != nullptr && range.end != nullptr,
+         "Setting a memory tag must be done within existing range");
+  if (range.start == nullptr || range.end == nullptr) {
+    log_debug(nmt)("set_tag called on non-existent range: " PTR_FORMAT " - " PTR_FORMAT,
+                   from, from+size);
+    return SummaryDiff();
+  }
+
+  StateType type = out_state(range.start).type();
+  RegionData new_data = RegionData(out_state(range.start).stack(), tag);
+
+  position end = MIN2(from + size, range.end->key());
+  SummaryDiff diff = register_mapping(from, end, type, new_data);
+  size = size - (end - from);
+  from = end;
+
+  // If end < from + sz then there are multiple ranges for which to set the flag.
+  while (end < from + size) {
+    range = _tree.find_enclosing_range(from);
```

Comment:
This is a bit of a brain teaser.

I first wondered why we needed to find the "from" node again, since we already have it. But the last `register_mapping` call can actually have coalesced ranges and either the old `from` node or the old `to` node can have disappeared. 

So, we have two possibilities: 
- the old "to" node is still there, in which case we could just read its RegionData state and use that for the next segment (modulo tag)
- the old "to" node disappeared due to merging (meaning, tagging the last segment causes it to coalesce with the next segment), in which case there is nothing to do.

But I am not 100% sure. So, re-querying the range from the treap is probably the safest way to go. 

But can you please add a comment stating that register_mapping may have modified the tree structure and caused segments to coalesce, just invalidating the old `range` ?

Reply:
Comment added.